src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,165 @@
+            add(value);
+        } else if (index == 0) {
+            addToHead(value);
+        } else {
+            Node<T> source = getNodeByIndex(index);
+            Node<T> toAdd = new Node<>(source.previous, value, source);
+            source.previous.next = toAdd;
+            source.previous = toAdd;
+            count++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexToGet(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexToGet(index);
+        Node<T> toSet = getNodeByIndex(index);
+        T oldValue = toSet.value;
+        toSet.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexToGet(index);
+        Node<T> toRemove = getNodeByIndex(index);
+        unlink(toRemove);
+        return toRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        return unlink(getNodeByValue(object));
     }
 
     @Override
     public int size() {
-        return 0;
+        return count;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    private void checkIndexToGet(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Invalid index " + index + " for size " + size());
+        }
+    }
+
+    private void checkIndexToAdd(int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException(
+                    String.format("Index %d out of bound for size %d", index, size()));
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        return index <= size() / 2 ? getNodeFromHead(index) :
+                getNodeFromTail(index);
+    }
+
+    private Node<T> getNodeFromTail(int index) {
+        int backCounter = size();
+        Node<T> current = tail;
+        while (--backCounter > index) {
+            current = current.previous;
+        }
+        return current;
+    }
+
+    private Node<T> getNodeFromHead(int index) {
+        int counter = 0;
+        Node<T> current = head;
+        while (counter++ < index) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private Node<T> getNodeByValue(T value) {
+        Node<T> current = head;
+        while (current != null) {
+            if (current.value == value || current.value != null
+                    && current.value.equals(value)) {
+                break;
+            }
+            current = current.next;
+        }
+        return current;
+    }
+
+    private boolean unlink(Node<T> node) {
+        if (node == null) {
+            return false;
+        }
```

Comment:
due to the comments regarding `containsNode()` method - you don't need to perform this check here after applying those changes